### PR TITLE
Refactor most of the code and stop using temp files

### PR DESCRIPTION
**Pull request by dominik-korsa**

### CODE REVIEW

**Review by MikolajKolek — CHANGES_REQUESTED** [review]
Overall, besides the changes I requested, it's looking pretty good, however I have some minor things you could change:
- After the refactor there are a couple of unused libraries in `Cargo.toml` that you could remove (like `lazy_static` for example)
- Toster seems to use kilobytes in its internal naming for memory usage variables, which is incorrect, as sio2jail returns memory usage in kibibytes
- The `expect` error naming scheme is very inconsistent - sometimes there is an exclamation mark at the end of the error, sometimes there isn't. In some places the different options are even used right side-by-side:
image
 I would suggest just removing all the exclamation marks, that would make it consistent

File: src/prepare_input.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            Test {
+                test_name,
+                input_source: TestInputSource::File(file_path)
+            }
+        })
+        .collect();
+
+    if tests.is_empty() {
+        println!("{}", "There are no files in the input directory with the provided file extension".red());
+        todo!("Implement error handling for reading inputs");
```
[the remark this answers is not in the document]
Comment (MikolajKolek) [reply src/prepare_input.rs]: This was fixed in the latest commit

File: src/test_errors.rs [hunk 2]
```diff
@@ -1,95 +1,78 @@
...
+	pub(crate) memory_kilobytes: Option<u64>,
+	pub(crate) time: Option<Duration>,
 }
 
-pub enum TestResult {
-	Correct {
-		test_name: String
-	},
+impl ExecutionMetrics {
+	#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (dominik-korsa) [reply src/test_errors.rs:12]: It is dead code in builds without sio2jail

File: src/executor/mod.rs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+pub(crate) mod simple;
+#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
+pub(crate) mod sio2jail;
+
+use std::process::Stdio;
+use crate::test_errors::{ExecutionError, ExecutionMetrics};
+
+pub(crate) trait TestExecutor: Sync + Send {
+    fn test_to_string(&self, input_stdio: Stdio) -> (ExecutionMetrics, Result<String, ExecutionError>);
```
[the remark this answers is not in the document]
Comment (dominik-korsa) [reply src/executor/mod.rs:9]: This was initially used to differentiate from the version writing output to file

**Review by MikolajKolek — APPROVED** [review]
Alright, looks good

**Pull request merged**